graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -69,76 +69,174 @@ def _get_last_edge_direction_and_name_to_location(location):
     _, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
-    # TODO(evan): If edge is recursed over, we need a more detailed statistic
     edge_counts = statistics.get_class_count(edge_name)
 
-    # Scale edge_counts if child_location's type is a subclass of the edge's endpoint type.
-    parent_name_from_edge, child_name_from_edge = _get_parent_and_child_name_from_edge(
-        schema_graph, child_location
-    )
+    parent_name_from_location = query_metadata.get_location_info(parent_location).type.name
     child_name_from_location = query_metadata.get_location_info(child_location).type.name
-    if child_name_from_edge != child_name_from_location:
-        # False-positive bug in pylint: https://github.com/PyCQA/pylint/issues/3039
-        # pylint: disable=old-division
+    parent_base_class_name, child_base_class_name = \
+        _get_base_class_names_of_parent_and_child_from_edge(schema_graph, child_location)
+
+    # False-positive bug in pylint: https://github.com/PyCQA/pylint/issues/3039
+    # pylint: disable=old-division
+    #
+    # Scale edge_counts if child_location's type is a subclass of the edge's endpoint type.
+    if child_name_from_location != child_base_class_name:
         edge_counts *= (
             float(statistics.get_class_count(child_name_from_location)) /
-            statistics.get_class_count(child_name_from_edge)
+            statistics.get_class_count(child_base_class_name)
+        )
+    # Scale edge_counts if parent_location's type is a subclass of the edge's endpoint type.
+    if parent_name_from_location != parent_base_class_name:
+        edge_counts *= (
+            float(statistics.get_class_count(parent_name_from_location)) /
+            statistics.get_class_count(parent_base_class_name)
+        )
+    # pylint: enable=old-division
+
+    return edge_counts
+
+
+def _estimate_edges_to_children_per_parent(
+    schema_graph, statistics, query_metadata, parameters, parent_location, child_location
+):
+    """Estimate the count of edges per parent_location that connect to child_location vertices.
+
+    Given a parent location of type A and child location of type B, assume all AB edges are
+    distributed evenly over A vertices, so the expected number of child edges per parent vertex is
+    (# of AB edges) / (# of A vertices).
+
+    Args:
+        schema_graph: SchemaGraph object.
+        statistics: Statistics object.
+        query_metadata: QueryMetadataTable object.
+        parameters: dict, parameters with which query will be executed.
+        parent_location: BaseLocation, corresponding to the location the edge traversal begins from.
+        child_location: BaseLocation, child of parent_location corresponding to the location the
+                        edge traversal ends at.
+
+    Returns:
+        float, expected number of edges per parent_location vertex that connect to child_location
+               vertices.
```

Comment:
nit: I'd indent this to under `float` instead of under `expected`, since this section does not have names followed by a `:`